libraries/AP_GPS/AP_GPS_MAV.cpp

```diff
@@ -42,67 +42,107 @@ bool AP_GPS_MAV::read(void)
-    loc.lat = packet.lat;
-    loc.lng = packet.lon;
-    if (have_alt) {
-        loc.alt = packet.alt;
+    switch (msg->msgid) {
+
+        case MAVLINK_MSG_ID_GPS_INPUT: {
+            mavlink_gps_input_t packet;
+            mavlink_msg_gps_input_decode(msg, &packet);
+
+            bool have_alt    = ((packet.ignore_flags & GPS_INPUT_IGNORE_FLAG_ALT) == 0);
+            bool have_hdop   = ((packet.ignore_flags & GPS_INPUT_IGNORE_FLAG_HDOP) == 0);
+            bool have_vdop   = ((packet.ignore_flags & GPS_INPUT_IGNORE_FLAG_VDOP) == 0);
+            bool have_vel_h  = ((packet.ignore_flags & GPS_INPUT_IGNORE_FLAG_VEL_HORIZ) == 0);
+            bool have_vel_v  = ((packet.ignore_flags & GPS_INPUT_IGNORE_FLAG_VEL_VERT) == 0);
+            bool have_sa     = ((packet.ignore_flags & GPS_INPUT_IGNORE_FLAG_SPEED_ACCURACY) == 0);
+            bool have_ha     = ((packet.ignore_flags & GPS_INPUT_IGNORE_FLAG_HORIZONTAL_ACCURACY) == 0);
+            bool have_va     = ((packet.ignore_flags & GPS_INPUT_IGNORE_FLAG_VERTICAL_ACCURACY) == 0);
+
+            state.time_week     = packet.time_week;
+            state.time_week_ms  = packet.time_week_ms;
+            state.status = (AP_GPS::GPS_Status)packet.fix_type;
+
+            Location loc = {};
+            loc.lat = packet.lat;
+            loc.lng = packet.lon;
+            if (have_alt) {
+                loc.alt = packet.alt;
+            }
+            state.location = loc;
+            state.location.options = 0;
+
+            if (have_hdop) {
+                state.hdop = packet.hdop * 100; // convert to centimeters
+            }
+
+            if (have_vdop) {
+                state.vdop = packet.vdop * 100; // convert to centimeters
+            }
+
+            if (have_vel_h) {
+                Vector3f vel(packet.vn, packet.ve, 0);
+                if (have_vel_v)
+                    vel.z = packet.vd;
+
+                state.velocity = vel;
+                state.ground_course = wrap_360(degrees(atan2f(vel.y, vel.x)));
+                state.ground_speed = norm(vel.x, vel.y);
+            }
+
+            if (have_sa) {
+                state.speed_accuracy = packet.speed_accuracy;
+                state.have_speed_accuracy = 1;
+            }
+
+            if (have_ha) {
+                state.horizontal_accuracy = packet.horiz_accuracy;
+                state.have_horizontal_accuracy = 1;
+            }
+
+            if (have_va) {
+                state.vertical_accuracy = packet.vert_accuracy;
+                state.have_vertical_accuracy = 1;
+            }
+
+            state.num_sats = packet.satellites_visible;
+            state.last_gps_time_ms = AP_HAL::millis();
+            _new_data = true;
+            break;
+            }
+
+        case MAVLINK_MSG_ID_HIL_GPS: {
+            mavlink_hil_gps_t packet;
+            mavlink_msg_hil_gps_decode(msg, &packet);
+
+            state.time_week = 0;
+            state.time_week_ms  = packet.time_usec/1000;
```

Comment:
If this doesn't increment in fixed/even steps I think that could be a problem for the EKF, (IE if it was 200ms then 180ms then 221 ms etc). Not sure what the sender code looks like, presumably its just locking GPS readings and passing them as they occurred so it would always count evenly.